### PR TITLE
Start MIP rare disease RNA through cg

### PR DESCRIPTION
**Pull request by ingkebil**
This PR adds rd rna pipeline start to `cg`.

**Test case 1**:
1. install on stage of the coffee machine: `bash servers/resources/hasta.scilifelab.se/update-cg-stage.sh mip-rdrna-pedigree`
1. install the config on stage: `bash servers/resources/hasta.scilifelab.se/update-servers-stage.sh mip-rd-rna`
1. install trailblazer on stage: `bash servers/resources/hasta.scilifelab.se/update-trailblazer-stage.sh mip-rd-rna`
1. activate stage: `us`
1. run: `cg analysis rna case-config poeticsnipe`
1. run: `cg analysis rna start --dry poeticsnipe`
1. run:
`cg analysis link -f poeticsnipe`
`cg analysis rna start poeticsnipe`
1. run: `trailblazer scan`

Test case 2: We also added an alias for the `config` command under `cg analysis`. Let's test if that still works:
1. run: `cg analysis config vitalmouse`
1. run: `cg analysis case-config vitalmouse`
1. run: `cg analysis config poeticsnipe`

**Expected outcome**:
1. creating the pedigree should create a pedigree here: `saved config to: /home/proj/stage/rare-disease/cases/vitalmouse/pedigree.yaml`
1. running `dry` will only print the `mip` command
1. running `start` will launch mip successfully!
  1. the case should show up on https://trailblazer-stage.scilifelab.se after starting it. The case should be marked as pending.
  1. Run `trailblazer scan` after the first job finishes: the case should be marked as running or failed.

Expected outcome test case 2:
1. Running both first commands should create a pedigree in `/home/proj/stage/rare-disease/cases/vitalmouse/pedigree.yaml`
1. Running poeticsnipe as a normal MIP case will error.

**Review:**
- [x] code approved by @barrystokman  @patrikgrenfeldt 
- [x] tests executed by @ingkebil 
- [x] "Merge and deploy" approved by @ingkebil 
Thanks for filling in who performed the code review and the test!

This is minor **version bump** because we're adding new functionality in a backwards compatible manner.

**post actions**
- [ ] Don't forget to add an RNA sample to the list of validation samples!
- [x] Don't forget to add the production config for mip7-rdrna in the servers repo @ingkebil 
- [ ] Don't forget to write a US about how to change validation of sampleInfo files based on ConfigSchema (see how we validate a pedigree in tb)

### CODE REVIEW

**Comment by ingkebil** [comment]
@emilia can you point me to an RNA case?

**Comment by ingkebil** [comment]
@patrikgrenfeldt the PR is still a draft, haven't tested this yet

**Comment by patrikgrenfeldt** [comment]
Ok, let's see where it heads then

**Comment by patrikgrenfeldt** [comment]
> @Emilia can you point me to an RNA case?

https://clinical-api-stage.scilifelab.se/admin/sample/?sort=4&desc=1&flt0_12=rna

**Comment by emiliaol** [comment]
Try `poeticsnipe`

File: cg/apps/mip/base.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    def build_command(self, config, case, **kwargs):
+        """Builds the command to execute MIP."""
+        command = [self.script, self.pipeline, case, CLI_OPTIONS['config']['option'], config]
+        for key, value in kwargs.items():
+            # enable passing in flags as "False" - shouldn't add command
+            if value:
+                command.append(CLI_OPTIONS[key]['option'])
+
+                # append value for non-flags
```
Comment (ingkebil) [inline cg/apps/mip/base.py]: weird looking code. Does this even work?

File: cg/cli/miprdrna.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    if case_obj is None:
+        LOGGER.error("%s: case not found", case_id)
+        context.abort()
+    if tb_api.analyses(family=case_obj.internal_id, temp=True).first():
+        LOGGER.warning("%s: analysis already running", case_obj.internal_id)
+    else:
+        email = email or environ_email()
+        kwargs = dict(config=context.obj['mip-rd-rna']['mip_config'], case=case_id,
+                      priority=priority, email=email, dryrun=dry, start_with=start_with)
+        if skip_evaluation:
```
Comment (ingkebil) [inline cg/cli/miprdrna.py]: @jemten is this still functional in MIP rd rna?

File: cg/apps/tb/api.py [hunk 3]
```diff
@@ -45,6 +45,14 @@ def start(self, case_id: str, priority: str = 'normal', email: str = None,
         self.commit()
         self.add_pending(case_id, email=email)
 
+    def mark_analysis_pending(self, case_id: str, email: str = None):
+        """ Add a new analysis to TB """
+        email = email or environ_email()
+        for old_analysis in self.analyses(family=case_id):
```
Comment (ingkebil) [inline cg/apps/tb/api.py]: split up into `delete_old_analyses`

File: cg/cli/miprdrna.py (hunk 2 above)
Comment (jemten) [reply cg/cli/miprdrna.py]: Nope. We haven't added the QC module yet so there are no evaluation to skip.

File: tests/apps/mip/conftest.py [hunk 4]
```diff
@@ -126,3 +128,10 @@ def tb_api():
     """Trailblazer API fixture"""
 
     return MockTB()
+
+
+@pytest.fixture(scope='session')
+def mip_api():
+    """MipAPI ficture"""
```
Comment (ingkebil) [inline tests/apps/mip/conftest.py]:
```suggestion
    """MipAPI fixture"""
```

File: tests/apps/mip/test_apps_mip_base.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+""" Test MIP app functionality """
+
+
+def test_build_command(mip_api):
+    """ Test building a command """
+    # GIVEN just a config
+    mip_config = 'tests/fixtures/global_config.yaml'
```
Comment (ingkebil) [inline tests/apps/mip/test_apps_mip_base.py]: move to fictures ;)

File: tests/cli/analysis/rna/conftest.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+class MockTB:
+    """Trailblazer mock fixture"""
+
+    def __init__(self):
+        self._link_was_called = False
+        self._family = None
+        self._temp = None
+
+    def link(self, family: str, sample: str, analysis_type: str, files: List[str]):
```
Comment (ingkebil) [inline tests/cli/analysis/rna/conftest.py]: remove!

File: tests/cli/analysis/rna/conftest.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        """Check if link has been called"""
+        return self._link_was_called
+
+    def analyses(self, family, temp):
+        """Mock TB analyses models"""
+
+        self._family = family
+        self._temp = temp
+
+        class Row:
```
Comment (ingkebil) [inline tests/cli/analysis/rna/conftest.py]: might be replaceable with a namedyuple

File: tests/cli/analysis/rna/conftest.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+
+@pytest.fixture(scope='function')
+def tb_api():
+    """Trailblazer API fixture"""
+
+    return MockTB()
+
+
+class MockStore():
```
Comment (ingkebil) [inline tests/cli/analysis/rna/conftest.py]: might be a namedtuple

File: tests/cli/analysis/rna/conftest.py (hunk 7 above)
Comment (ingkebil) [reply tests/cli/analysis/rna/conftest.py]: tried, doesn't work

File: tests/cli/analysis/rna/conftest.py (hunk 8 above)
Comment (ingkebil) [reply tests/cli/analysis/rna/conftest.py]: doesn't work

**Comment by ingkebil** [comment]
poeticsnipe has no expected_coverage set :(

**Comment by ingkebil** [comment]
@emiliaol poeticsnipe has no expected_coverage set :(

**Comment by ingkebil** [comment]
Testing:

```
[0|0|3001] 1z [kenny.billiau@hasta:~/git/trailblazer] [D_cg_190715_KB] master(+1/-1) 11s ± cg analysis rna case-config poeticsnipe
setting 'unknown' phenotype to 'unaffected'
saved config to: /home/proj/stage/rare-disease/cases/poeticsnipe/pedigree.yaml
```

:heavy_check_mark:

**Comment by ingkebil** [comment]
testing dry run:
```
[0|0|3000] 1z [kenny.billiau@hasta:~/git/trailblazer] [D_cg_190715_KB] master(+1/-1) 17s ± cg analysis rna start --dry poeticsnipe
/home/proj/bin/conda/envs/S_mip7_rd-rna_191001/bin/mip analyse rd_rna poeticsnipe --config_file /home/kenny.billiau/git/servers/config/hasta.scilifelab.se/mip7-rd-rna-stage.yaml --email kenny.billiau@scilifelab.se --dry_run_all 1
```

**Comment by ingkebil** [comment]
Testing again after changes in TB:

```
[0|0|2714] 1z [kenny.billiau@hasta:~/git/trailblazer] [D_cg_190715_KB] mip-rd-rna(+33/-4) 13s 148 ± cg analysis rna case-config poeticsnipe
setting 'unknown' phenotype to 'unaffected'
saved config to: /home/proj/stage/rare-disease/cases/poeticsnipe/pedigree.yaml
```

:heavy_check_mark:

**Comment by ingkebil** [comment]
```
[0|0|2733] 1z [kenny.billiau@hasta:~/git/cg] [D_cg_190715_KB] mip-rdrna-pedigree 4s ± cg analysis rna start --dry poeticsnipe
/home/proj/bin/conda/envs/S_mip7_rd-rna_191001/bin/mip analyse rd_rna poeticsnipe --config_file /home/kenny.billiau/git/servers/config/hasta.scilifelab.se/mip7-rd-rna-stage.yaml --email kenny.billiau@scilifelab.se --dry_run_all 1
```

:heavy_check_mark:

**Comment by ingkebil** [comment]
marked as pending!

Screenshot from 2019-10-09 14-49-58

**Comment by ingkebil** [comment]
```
[1|31|2731] 1z [kenny.billiau@hasta:~/git/trailblazer] [D_cg_190715_KB] mip-rd-rna(1) ± cg analysis config vitalmouse                                                                                                                                                                       
saved config to: /home/proj/stage/rare-disease/cases/vitalmouse/pedigree.yaml
[1|31|2756] 1z [kenny.billiau@hasta:~/git/trailblazer] [D_cg_190715_KB] mip-rd-rna(1) 22s 1 ± cg analysis case-config vitalmouse
saved config to: /home/proj/stage/rare-disease/cases/vitalmouse/pedigree.yaml
[1|31|2756] θ62° 1z [kenny.billiau@hasta:~/git/trailblazer] [D_cg_190715_KB] mip-rd-rna(1) 21s ± cg analysis config poeticsnipe                                                                                                                                                             
ACC5905A2 -> analysis_type: Not a valid choice.
Traceback (most recent call last):
  File "/home/proj/bin/conda/envs/D_cg_190715_KB/bin/cg", line 11, in <module>
    load_entry_point('cg', 'console_scripts', 'cg')()
  File "/home/proj/bin/conda/envs/D_cg_190715_KB/lib/python3.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/proj/bin/conda/envs/D_cg_190715_KB/lib/python3.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/proj/bin/conda/envs/D_cg_190715_KB/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/bin/conda/envs/D_cg_190715_KB/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/bin/conda/envs/D_cg_190715_KB/lib/python3.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/proj/bin/conda/envs/D_cg_190715_KB/lib/python3.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/proj/bin/conda/envs/D_cg_190715_KB/lib/python3.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/kenny.billiau/git/cg/cg/cli/analysis.py", line 84, in case_config
    config_data = context.obj['api'].config(family_obj)
  File "/home/kenny.billiau/git/cg/cg/meta/analysis.py", line 108, in config
    config_data = self.tb.make_config(data, pipeline)
  File "/home/kenny.billiau/git/trailblazer/trailblazer/mip/config.py", line 71, in make_config
    self.validate_config(data, pipeline)
  File "/home/kenny.billiau/git/trailblazer/trailblazer/mip/config.py", line 91, in validate_config
    raise ConfigError('invalid config input', errors=errors)
trailblazer.exc.ConfigError: invalid config input
```

**Comment by ingkebil** [comment]
Testing `trailblazer scan`

```
[0|0|2690] θ61° [kenny.billiau@hasta:~/git/trailblazer] [D_cg_190715_KB] mip-rd-rna ± trailblazer scan
unexpected output, missing key: case_id in <_io.TextIOWrapper name='/home/proj/stage/rare-disease/cases/grownliger/analysis/grownliger_config.yaml' mode='r' encoding='UTF-8'>
unexpected output, missing key: analysis_date in <_io.TextIOWrapper name='/home/proj/stage/rare-disease/cases/15138/analysis/15138_config.yaml' mode='r' encoding='UTF-8'>
/home/proj/bin/conda/envs/D_cg_190715_KB/lib/python3.7/site-packages/pymysql/cursors.py:170: Warning: (1265, "Data truncated for column 'type' at row 1")
  result = self._query(query)
New log added: poeticsnipe (129454) - failed
```

Log added! :)

:heavy_check_mark:

File: tests/cli/analysis/rna/conftest.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    """We need to call the family function from the store
+    without accessing the database. So here we go"""
+
+    def __init__(self):
+        """Constructor"""
```
Comment (patrikgrenfeldt) [inline tests/cli/analysis/rna/conftest.py]: yeah?

File: tests/cli/analysis/rna/test_cli_analysis_rna.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    # THEN the command should be printed
+    with caplog.at_level(logging.INFO):
+        assert '${HOME}/bin/mip analyse rd_rna angrybird --config_file config.yaml ' \
+               '--email james.holden@scilifelab.se --dry_run_all' in caplog.text
+
+
+def test_start(cli_runner, tb_api, mock_store, caplog, monkeypatch):
+    """Test starting MIP"""
+
+    # pylint: disable=unused-argument
```
Comment (patrikgrenfeldt) [inline tests/cli/analysis/rna/test_cli_analysis_rna.py]: I'm no fan of silencing the linter, could we delete the arguments instead, e.g. "del kwargs"

Comment (ingkebil) [reply tests/cli/analysis/rna/test_cli_analysis_rna.py]: sure!

File: tests/cli/analysis/rna/conftest.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+                """Mock constructor"""
+
+                self._first_was_called = False
+
+            def first(self):
+                """Mock that the first row doesn't exist"""
+
+                self._first_was_called = True
+
+                return False
```
[the remark this answers is not in the document]
Comment (ingkebil) [reply tests/cli/analysis/rna/conftest.py]: According to the sqlalchemy docs, that is what `first()` returns if the set is empty: https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.first

**Pull request merged**